### PR TITLE
rectangular borders and cursor for minimap

### PR DESCRIPTION
**Pull request by Incompleteusern**
image

Code supports custom map sizes too:
image

Looking for feedback on code quality

### CODE REVIEW

File: common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java [hunk 1]
```diff
@@ -136,6 +142,80 @@ private void rotateMapToPlayer(PoseStack poseStack, float centerX, float centerZ
...
+        // TODO move most of the buffer builder code into RenderUtils, maybe separate each map border into different
+        // files to avoid insanity involving tx1/ty1/tx2/ty2 + remove magic constant
+        private void renderRectangularMapBorder(
+                PoseStack poseStack, float renderX, float renderY, float width, float height) {
+            Texture texture = null;
+            int grooves = 0;
+            int tx1 = 0;
+            int ty1 = 0;
+            int tx2 = 0;
+            int ty2 = 0;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java]: MapBorderType could have these?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java]: At least that is how we do this with health bar overlay

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java]: Mhm that'd be better

File: common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java [hunk 2]
```diff
@@ -136,6 +142,80 @@ private void rotateMapToPlayer(PoseStack poseStack, float centerX, float centerZ
...
+            bufferBuilder
+                    .vertex(matrix, (renderX + width + groovesWidth), (renderY - groovesHeight), 0)
+                    .uv((tx2) * uScale, (ty1) * vScale)
+                    .endVertex();
+            bufferBuilder
+                    .vertex(matrix, (renderX - groovesWidth), (renderY - groovesHeight), 0)
+                    .uv((tx1) * uScale, (ty1) * vScale)
+                    .endVertex();
+            bufferBuilder.end();
+            BufferUploader.end(bufferBuilder);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java]: What is the difference between this and RenderUtils#drawTexturedRect?

File: common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java [hunk 3]
```diff
@@ -106,17 +110,19 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
                     // case Circle -> {
                     // TODO
                     // }
             }
 
             // TODO cursor
 
-            // TODO Border rendering
+            // render border
+            switch (maskType) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java:143]: Isn't it a bit weird to have 3 switches of the same thing?

File: common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java (hunk 2 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java]: This specifies `tx1` and `tx2`, which is not possible with `RenderUtils#drawTexturedRect` right now (outside of changing textureWidth and height)

File: common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java (hunk 3 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java:143]: I don't see an alternative as of now, but I can replace it with if statements if desired

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java:143]: I am talking about only having 1 switch and doing everything needed for rendering setup there.

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java:143]: I think we need at least 2 switches as they need to enable and disable the mask, but the last two maybe can be merged a bit clunkily.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/MiniMapOverlayFeature.java:143]: Ah. Okay.

File: common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java [hunk 4]
```diff
@@ -136,6 +140,55 @@ private void rotateMapToPlayer(PoseStack poseStack, float centerX, float centerZ
             poseStack.translate(-centerX, -centerZ, 0);
         }
 
+        // TODO move most of the buffer builder code into RenderUtils, maybe separate each map border into different
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java]: We can remove this todo by just making a method for this in RenderUtils? The other part of todo is already done

File: common/src/main/java/com/wynntils/mc/render/Texture.java [hunk 5]
```diff
@@ -16,7 +16,11 @@ public enum Texture {
     LOOTRUN_LINE("path_arrow.png", 16, 16),
     MANA_BAR("bars_mana.png", 256, 256),
     OVERLAY_SELECTION_GUI("overlay_selection_gui.png", 195, 256),
-    WYNNCRAFT_ICON("wynncraft_icon.png", 64, 64);
+    WYNNCRAFT_ICON("wynncraft_icon.png", 64, 64),
+
+    WYNN_MAP_TEXTURES("map/wynn_map_textures.png", 256, 256),
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/render/Texture.java]: Could you order this alphabetically? (and perhaps have a comment after newline (`//map textures`?

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/render/Texture.java]: Will do

**Comment by Incompleteusern** [comment]
image

File: common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java [hunk 6]
```diff
@@ -104,38 +113,100 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
-            // TODO Border rendering
-
             // TODO Directional Text
 
             // TODO Coords
 
         }
 
-        private void rotateMapToPlayer(PoseStack poseStack, float centerX, float centerZ) {
+        private void rotateMap(PoseStack poseStack, float centerX, float centerZ, float angle) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java]: Can we rename this? This does not actually rotate the map, it just does poseStack translations.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java]: Perhaps rotatePose?

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java]: I think I will move the method to `RenderUtils` as well

**Review by kristofbolyai — APPROVED** [review]
This looks okay for now, obviously there are a lot of todos in the code, so it is hard to judge the overall quality of the feature's code. I think this is nice enough progress to merge, and I know you are gonna finish this.

**Pull request merged**